network/kademlia_load_balancer_test.go

```diff
@@ -0,0 +1,338 @@
+	tk := newTestKademlia(t, "11111111")
+	klb := NewKademliaLoadBalancer(tk, false)
+	caps := make(map[string]*capability.Capability)
+
+	capKey := "42:101"
+	caps[capKey] = capability.NewCapability(42, 3)
+	_ = caps[capKey].Set(0)
+	_ = caps[capKey].Set(2)
+	_ = tk.RegisterCapabilityIndex(capKey, *caps[capKey])
+
+	capPeer := tk.newTestKadPeerWithCapabilities("10100000", caps[capKey])
+	tk.Kademlia.On(capPeer)
+	useStats := klb.resourceUseStats
+	useStats.WaitKey(capPeer.Key())
+	tk.On("01010101") // bin 0 dec 85 hex 55
+	useStats.WaitKey(bitStringToHex("01010101"))
+	tk.On("01010100") // bin 0 dec 84 hex 54
+	useStats.WaitKey(bitStringToHex("01010100"))
+	tk.On("10010100") // bin 1 dec 148
+	useStats.WaitKey(bitStringToHex("10010100"))
+	tk.On("10010001") // bin 1 dec 145
+	useStats.WaitKey(bitStringToHex("10010001"))
+	tk.On("11010100") // bin 2 dec 212
+	useStats.WaitKey(bitStringToHex("11010100"))
+	tk.On("11010101") // bin 2 dec 213
+	useStats.WaitKey(bitStringToHex("11010101"))
+	stats := make(map[string]int)
+	countUse := func(bin LBBin) bool {
+		peer := bin.LBPeers[0].Peer
+		bin.LBPeers[0].AddUseCount()
+		key := peerToBitString(peer)
+		stats[key] = stats[key] + 1
+		return false
+	}
+
+	pivotAddressBin1 := pot.NewAddressFromString("10000000") // Two nearest peers (1,2)
+	// Instead of selecting peers 10010100 or 10010001, capPeer is always chosen (10100000)
+	klb.EachBinFiltered(pivotAddressBin1, capKey, countUse)
+	klb.EachBinFiltered(pivotAddressBin1, capKey, countUse)
+	klb.EachBinFiltered(pivotAddressBin1, capKey, countUse)
+
+	count := useStats.GetUses(capPeer)
+	if count != 3 || stats["10100000"] != 3 {
+		t.Errorf("Expected 3 uses of capability peer but got %v/%v", count, stats["10100000"])
+	}
+
+	secondCapPeer := tk.newTestKadPeerWithCapabilities("10100001", caps[capKey])
+	tk.Kademlia.On(secondCapPeer)
+	useStats.WaitKey(secondCapPeer.Key())
+	secondCountStart := useStats.GetUses(secondCapPeer)
+	count = useStats.GetUses(capPeer)
+	klb.EachBinFiltered(pivotAddressBin1, capKey, countUse)
+	klb.EachBinFiltered(pivotAddressBin1, capKey, countUse)
+	secondCount := useStats.GetUses(secondCapPeer)
+	if secondCount-secondCountStart != 2 {
+		t.Errorf("Expected 2 uses of second capability peer but got %v", secondCount-secondCountStart)
+	}
+
+}
+
+// TestResourceUseStats checks that on and off messages are delivered in order
+func TestResourceUseStats(t *testing.T) {
+
+	testResourceUseStats := func(t *testing.T, delay time.Duration) {
+		k := NewKademlia(make([]byte, 32), NewKadParams())
+		lb := NewKademliaLoadBalancer(k, false)
+		for i := uint64(0); i < 10; i++ {
+			a := make([]byte, 8)
+			binary.BigEndian.PutUint64(a, i)
+			p := NewPeer(&BzzPeer{BzzAddr: NewBzzAddr(a, a)}, nil)
+			k.On(p)
+			if delay > 0 {
+				time.Sleep(delay)
+			}
+			k.Off(p)
+			if delay > 0 {
+				time.Sleep(delay)
+			}
+		}
+
+		// we need to sleep to allow all messages to be received by lb
+		time.Sleep(100 * time.Millisecond)
```

Comment:
If this is causing flakiness on Travis, we should replace it with a retry at least.